### PR TITLE
[Makefile/CI] Parallelize lint and unit test

### PR DESCRIPTION
**Pull request by djaglowski**
This change introduces a mechanism for grouping and parallelizing per-module make tasks. This should cut some notable time off of the CI process and give developers feedback sooner when a failure is detected. It also enables developers to run more targeted tests locally as well.

This PR parallelizes the `lint` and `unittest` CI jobs, which currently run for 15-30 minutes. With the change, run times on these jobs appear to stay below 10 minutes.

This change impacted the way that integration tests were run, so that target was updated as well. Modules containing integration tests previously needed to be hardcoded into the Makefile. This has repeatedly led to oversights, omitting new modules from integration tests. The new pattern is that any module containing a file named `*integration_test.go` or `*e2e_test.go` will be included in the target. This currently picks all integration tests in the repo (excluding testbed) and is less likely to be overlooked than the hardcoded list in the Makefile.

### CODE REVIEW

**Comment by djaglowski** [comment]
Currently includes unmerged PR #10031

File: Makefile [hunk 1]
```diff
@@ -11,37 +11,53 @@ BUILD_INFO=-ldflags "-X $(BUILD_INFO_IMPORT_PATH).Version=$(VERSION)"
...
-	receiver/nginxreceiver \
-	receiver/postgresqlreceiver \
-	receiver/redisreceiver \
-	receiver/riakreceiver \
-	receiver/zookeeperreceiver \
-	extension/observer/dockerobserver
+GROUP ?= all
+FOR_GROUP_TARGET=for-$(GROUP)-target
+
+TEMP_EX_STANZA=-not -path "./pkg/stanza/*"
```
Comment (djaglowski) [inline Makefile:17]: This can be removed if #10013 is merged

File: Makefile [hunk 2]
```diff
@@ -11,37 +11,53 @@ BUILD_INFO=-ldflags "-X $(BUILD_INFO_IMPORT_PATH).Version=$(VERSION)"
...
+
+RECEIVER_MODS_0 := $(shell find ./receiver/[a-k]* $(FIND_MOD_ARGS) -exec $(TO_MOD_DIR) )
+RECEIVER_MODS_1 := $(shell find ./receiver/[l-z]* $(FIND_MOD_ARGS) -exec $(TO_MOD_DIR) )
+RECEIVER_MODS := $(RECEIVER_MODS_0) $(RECEIVER_MODS_1)
+PROCESSOR_MODS := $(shell find ./processor/* $(FIND_MOD_ARGS) -exec $(TO_MOD_DIR) )
+EXPORTER_MODS := $(shell find ./exporter/* $(FIND_MOD_ARGS) -exec $(TO_MOD_DIR) )
+EXTENSION_MODS := $(shell find ./extension/* $(FIND_MOD_ARGS) -exec $(TO_MOD_DIR) )
+INTERNAL_MODS := $(shell find ./internal/* $(FIND_MOD_ARGS) -exec $(TO_MOD_DIR) )
+OTHER_MODS := $(shell find . $(EX_COMPONENTS) $(EX_INTERNAL) $(FIND_MOD_ARGS) -exec $(TO_MOD_DIR) ) $(PWD)
+ALL_MODS := $(RECEIVER_MODS) $(PROCESSOR_MODS) $(EXPORTER_MODS) $(EXTENSION_MODS) $(INTERNAL_MODS) $(OTHER_MODS)
```
Comment (TylerHelmuth) [inline Makefile:34]: Would it make sense to spilt processors and/or exporters and/or extensions into multiple groups as well?  I know there aren't as many of those as there are receivers right now, but I'd expect them to continue to grow. Would be great for the CI to be able to handle increases in scale without having to come back and update this process again.

Comment (djaglowski) [reply Makefile:34]: In my opinion we're better off parallelizing only to the extent that it benefits the current set of components. If we end up seeing that splitting another group would be helpful, it's it fairly easy thing to do, given the pattern established here.

Comment (TylerHelmuth) [reply Makefile:34]: Is there a cost to the parallelization?

**Review by TylerHelmuth — APPROVED** [review]

**Review by codeboten — APPROVED** [review]

**Pull request merged**